README.md
# Unity Asynchronous Image Loader

[`ImageConversion.LoadImage`](https://docs.unity3d.com/ScriptReference/ImageConversion.LoadImage.html) and `Texture2D.LoadImage` are slow when loading large images (greater than 2K) at runtime. They blocks the Unity main thread when loading the image for a duration between a hundred milliseconds and even a few seconds. This is a dealbreaker for those games and applications that want to load those images programmatically at runtime.

This package aims to offload image loading, image decoding and mipmap generation to other threads. It creates smoother gameplay and reduces lag spike on the Unity main thread when loading large images.

This package uses [FreeImage](https://freeimage.sourceforge.io/). which is the same library used by Unity to process image data.

## Unity Version

This package is developed in Unity 2019.1. It may work on Unity 2020 and Unity 2021.

## Installation

The package can be installed using the Git URL `https://github.com/Looooong/UnityAsyncImageLoader.git` by following [Installing from a Git URL](https://docs.unity3d.com/Manual/upm-ui-giturl.html) instructions.

## Dependencies

+ Unity Burst
+ Unity Mathematics

## Usage

### Loader Settings

```cs
  /// <summary>Settings used by the image loader.</summary>
  public struct LoaderSettings {
    /// <summary>Create linear texture. Only applicable to methods that create new <c>Texture2D</c>. Defaults to false.</summary>
    public bool linear;
    /// <summary>Texture data won't be readable on the CPU after loading. Defaults to false.</summary>
    public bool markNonReadable;
    /// <summary>Whether or not to generate mipmaps. Defaults to true.</summary>
    public bool generateMipmap;
    /// <summary>Automatically calculate the number of mipmap levels. Defaults to true. Only applicable to methods that create new <c>Texture2D</c>.</summary>
    public bool autoMipmapCount;
    /// <summary>Mipmap count, including the base level. Must be greater than 1. Only applicable to methods that create new <c>Texture2D</c>.</summary>
    public int mipmapCount;
    /// <summary>Used to explicitly specify the image format. Defaults to FIF_UNKNOWN, which the image format will be automatically determined.</summary>
    public FreeImage.Format format;
    /// <summary>Whether or not to log exception caught by this method. Defaults to true.</summary>
    public bool logException;

    public static LoaderSettings Default => new LoaderSettings {
      linear = false,
      markNonReadable = false,
      generateMipmap = true,
      autoMipmapCount = true,
      format = FreeImage.Format.FIF_UNKNOWN,
      logException = true,
    };
  }
```

### Load Image Asynchronously

```cs
  var imageData = File.ReadAllBytes();
  var texture = new Texture2D(1, 1);
  var loaderSettings = AsyncImageLoader.LoaderSettings.Default;
  var success = false;

  // =====================================
  // Load image data into existing texture
  // =====================================

  // Use the default LoaderSettings
  success = await AsyncImageLoader.LoadImageAsync(texture, imageData);

  // Similar to ImageConversion.LoadImage
  // Mark texture as unreadable after reading.
  success = await AsyncImageLoader.LoadImageAsync(texture, imageData, true);

  // Use a custom LoaderSettings
  success = await AsyncImageLoader.LoadImageAsync(texture, imageData, loaderSettings);

  // ==================================
  // Create new texture from image data
  // ==================================

  // Use the default LoaderSettings
  texture = await AsyncImageLoader.CreateFromImageAsync(imageData);

  // Use a custom LoaderSettings
  texture = await AsyncImageLoader.CreateFromImageAsync(imageData, loaderSettings);
```

### Load Image Synchronously

The synchronous variants are the same as the  asynchronous counterparts but without `Async` suffix in theirs name. They are useful for debugging and profiling within a frame.

```cs
  var imageData = File.ReadAllBytes();
  var texture = new Texture2D(1, 1);
  var loaderSettings = AsyncImageLoader.LoaderSettings.Default;
  var success = false;

  // =====================================
  // Load image data into existing texture
  // =====================================

  // Use the default LoaderSettings
  success = AsyncImageLoader.LoadImage(texture, imageData);

  // Similar to ImageConversion.LoadImage
  // Mark texture as unreadable after reading.
  success = AsyncImageLoader.LoadImage(texture, imageData, true);

  // Use a custom LoaderSettings
  success = AsyncImageLoader.LoadImage(texture, imageData, loaderSettings);

  // ==================================
  // Create new texture from image data
  // ==================================

  // Use the default LoaderSettings
  texture = AsyncImageLoader.CreateFromImage(imageData);

  // Use a custom LoaderSettings
  texture = AsyncImageLoader.CreateFromImage(imageData, loaderSettings);
```

## After Loading

### Texture Format

If the image has alpha channel, the format will be `RGBA32`, otherwise, it will be `RGB24`.

### Mipmap Count

If the `LoadImage` and `LoadImageAsync` variants are used with `generateMipmap` set to `true`, the mipmap count is set to the maximum possible number for a particular texture. If you want to control the number of mipmap, you can use the `CreateFromImage` and `CreateFromImageAsync` instead.

### Mipmap Data

The mipmaps are generated using box filtering with 2x2 kernel. The final result won't be the same as Unity's counterpart when using texture import in the editor.

## Troubleshooting

### There is still lag spike when loading large images

After `AsyncImageLoader` method finishes executing, the image data are still transfering to the GPU. Therefore, any object, like material or UI, that wants to use the texture afterward will have to wait for the texture to finish uploading and thus block the Unity main thread.

There is no easy way to detect if the texture has finished uploading its data. The workarounds are:
+ Wait for a second or more before using the texture.
+ **(Not tested)** Use [`AsyncGPUReadback`](https://docs.unity3d.com/ScriptReference/Rendering.AsyncGPUReadback.html) to request a single pixel from the texture. It will wait for the texture to finish uploading before downloading that single pixel. Then the request callback can be used to notify the Unity main thread about the texture upload completion.

## Acknowledgement

This package is inspired by Matias Lavik's [`unity-async-textureimport`](https://codeberg.org/matiaslavik/unity-async-textureimport).


CHANGELOG.md
## [Unreleased]

## [0.1.3] - 2022-12-19
### Enhancement
- Add support for non-bitmap image type.

### Fixed
- Fix accuracy loss when generating mipmap (#7).

## [0.1.2] - 2022-02-17
### Fixed
- Fix Burst complain about two containers maybe aliasing in FilterMipmapJob.
- Use `Texture2D.Reinitialize` instead of `Texture2D.Resize` from Unity 2021.1 onward.

## [0.1.1] - 2021-07-30
### Added
- Add `CHANGELOG.md`.

### Changed
- Refactor Burst jobs.

### Fixed
- Fix typo in `README.md`.
- Fix performance issue when importing PNG without alpha channel.
- Fix out of bound access in mipmap generation job.

## [0.1.0] - 2021-07-28
### Added
- Implement `AsyncImageLoader`.

[Unreleased]: https://github.com/Looooong/UnityAsyncImageLoader/compare/v0.1.3...HEAD
[0.1.3]: https://github.com/Looooong/UnityAsyncImageLoader/releases/tag/v0.1.3
[0.1.2]: https://github.com/Looooong/UnityAsyncImageLoader/releases/tag/v0.1.2
[0.1.1]: https://github.com/Looooong/UnityAsyncImageLoader/releases/tag/v0.1.1
[0.1.0]: https://github.com/Looooong/UnityAsyncImageLoader/releases/tag/v0.1.0


Tests/TestMipmap.cs
using NUnit.Framework;
using UnityEngine;

public class TestMipmap {
  [Test]
  public void TestPNGImage() {
    var data = new byte[] {
      0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A, 0x00, 0x00, 0x00, 0x0D, 0x49, 0x48, 0x44, 0x52,
      0x00, 0x00, 0x00, 0x02, 0x00, 0x00, 0x00, 0x02, 0x08, 0x06, 0x00, 0x00, 0x00, 0x72, 0xB6, 0x0D,
      0x24, 0x00, 0x00, 0x00, 0x1B, 0x49, 0x44, 0x41, 0x54, 0x08, 0xD7, 0x63, 0x60, 0xF8, 0xCF, 0xD0,
      0xC0, 0xC0, 0xF0, 0xDF, 0x81, 0xE1, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0x33, 0x30, 0xEC, 0x07, 0x00,
      0x43, 0x18, 0x08, 0x79, 0xEB, 0xE3, 0x55, 0xB4, 0x00, 0x00, 0x00, 0x00, 0x49, 0x45, 0x4E, 0x44,
      0xAE, 0x42, 0x60, 0x82
    };

    var texture = AsyncImageLoader.CreateFromImage(data);
    var mipmapLevel0 = texture.GetPixelData<int>(0);
    var mipmapLevel1 = texture.GetPixelData<int>(1);

    Assert.That(texture.format, Is.EqualTo(TextureFormat.RGBA32));
    Assert.That(texture.width, Is.EqualTo(2));
    Assert.That(texture.height, Is.EqualTo(2));
    Assert.That(texture.mipmapCount, Is.EqualTo(2));

    Assert.That(mipmapLevel0.Length, Is.EqualTo(4));
    Assert.That(mipmapLevel0, Is.EquivalentTo(new int[] {
      unchecked((int)0xFF_FF_FF_FF),
      unchecked((int)0xBF_00_00_FF),
      unchecked((int)0x80_00_FF_00),
      unchecked((int)0x40_FF_00_00),
    }));

    Assert.That(mipmapLevel1.Length, Is.EqualTo(1));
    Assert.That(mipmapLevel1, Is.EquivalentTo(new int[] {
      unchecked((int)0x9F_7F_7F_7F),
    }));
  }

  [Test]
  public void TestJPEGImage() {
    var data = new byte[] {
      0xFF, 0xD8, 0xFF, 0xE0, 0x00, 0x10, 0x4A, 0x46, 0x49, 0x46, 0x00, 0x01, 0x01, 0x01, 0x01, 0x2C,
      0x01, 0x2C, 0x00, 0x00, 0xFF, 0xDB, 0x00, 0x43, 0x00, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
      0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
      0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
      0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
      0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0xFF, 0xDB, 0x00, 0x43, 0x01, 0x01, 0x01,
      0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
      0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
      0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
      0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0xFF, 0xC9,
      0x00, 0x11, 0x08, 0x00, 0x02, 0x00, 0x02, 0x03, 0x01, 0x11, 0x00, 0x02, 0x11, 0x01, 0x03, 0x11,
      0x01, 0xFF, 0xCC, 0x00, 0x0A, 0x00, 0x10, 0x10, 0x05, 0x01, 0x10, 0x11, 0x05, 0xFF, 0xDA, 0x00,
      0x0C, 0x03, 0x01, 0x00, 0x02, 0x11, 0x03, 0x11, 0x00, 0x3F, 0x00, 0xFF, 0x00, 0xED, 0x83, 0x1F,
      0x33, 0xBE, 0xF8, 0x94, 0x15, 0x70, 0xB2, 0xBB, 0x79, 0x32, 0xF4, 0x33, 0x76, 0x95, 0xE5, 0x72,
      0x12, 0x80, 0xC7, 0xBB, 0xD4, 0xFB, 0xA3, 0x62, 0x4D, 0xF7, 0x60, 0x15, 0xAF, 0x52, 0x24, 0xCD,
      0xD2, 0xBF, 0xC1, 0x1C, 0xF3, 0x9A, 0x08, 0x87, 0xEE, 0x86, 0xB1, 0xD7, 0xE2, 0xC8, 0x9D, 0x24,
      0x55, 0xC1, 0x6D, 0xFF, 0x00, 0x4F, 0x49, 0x19, 0x7B, 0x42, 0xB3, 0x28, 0xB9, 0xB6, 0x05, 0x0E,
      0x5F, 0xCF, 0x84, 0x92, 0x6C, 0xB5, 0x54, 0x89, 0xBF, 0xFE, 0xE0, 0x0F, 0x33, 0x5E, 0x63, 0xB3,
      0x89, 0xEB, 0x11, 0x5A, 0x95, 0xD5, 0x09, 0x39, 0x2C, 0xB6, 0x04, 0x72, 0x3D, 0xF4, 0x9B, 0xB4,
      0x58, 0x49, 0x64, 0x46, 0x1D, 0x65, 0x60, 0x14, 0xBD, 0x04, 0x5C, 0xA4, 0xF0, 0x82, 0xAF, 0x10,
      0xC8, 0xFD, 0x60, 0x8E, 0x9C, 0x55, 0x8C, 0xA4, 0xFF, 0x00, 0x3C, 0xAE, 0x16, 0xBB, 0x53, 0xC2,
      0x59, 0x82, 0xDD, 0xAD, 0x3F, 0x7D, 0x68, 0x04, 0x6D, 0x7E, 0xD3, 0x6B, 0xA9, 0xA2, 0x01, 0x7F,
      0x6E, 0xA9, 0xBE, 0xE9, 0x3E, 0xB6, 0x24, 0x9E, 0xAB, 0xAD, 0xD8, 0xE7, 0x60, 0x64, 0x73, 0xC3,
      0x8F, 0xDC, 0xDC, 0xC2, 0x2F, 0x35, 0xF1, 0xEB, 0x2F, 0xA2, 0x75, 0x61, 0x01, 0xDF, 0x11, 0xF4,
      0xD0, 0xC1, 0x62, 0x5A, 0x28, 0x22, 0xEC, 0x05, 0x2A, 0x41, 0x14, 0xC2, 0xA0, 0x4D, 0xB7, 0x9F,
      0x3C, 0xF5, 0xD8, 0x90, 0xF1, 0x6F, 0x87, 0x09, 0xBE, 0xA9, 0xFD, 0x44, 0xBD, 0x59, 0x0C, 0x8B,
      0x03, 0xF8, 0xB1, 0x0E, 0x07, 0x73, 0x3C, 0x82, 0xDA, 0x09, 0x83, 0x35, 0x03, 0x90, 0x3E, 0xBC,
      0x65, 0x4B, 0x44, 0xAF, 0x2C, 0x80, 0xFF, 0xD9
    };

    var texture = AsyncImageLoader.CreateFromImage(data);
    var mipmapLevel0 = texture.GetPixelData<byte>(0);
    var mipmapLevel1 = texture.GetPixelData<byte>(1);

    Assert.That(texture.format, Is.EqualTo(TextureFormat.RGB24));
    Assert.That(texture.width, Is.EqualTo(2));
    Assert.That(texture.height, Is.EqualTo(2));
    Assert.That(texture.mipmapCount, Is.EqualTo(2));

    Assert.That(mipmapLevel0.Length, Is.EqualTo(3 * 4));
    Assert.That(mipmapLevel0, Is.EquivalentTo(new byte[] {
      0xFF, 0xFF, 0xFF,
      0xFE, 0x00, 0x00,
      0x00, 0xFF, 0x01,
      0x00, 0x00, 0xFE
    }));

    Assert.That(mipmapLevel1.Length, Is.EqualTo(3));
    Assert.That(mipmapLevel1, Is.EquivalentTo(new byte[] {
      0x7F, 0x7F, 0x7F,
    }));
  }

  [Test]
  public void TestEXRImage() {
    var data = new byte[] {
      0x76, 0x2F, 0x31, 0x01, 0x02, 0x00, 0x00, 0x00, 0x63, 0x68, 0x61, 0x6E, 0x6E, 0x65, 0x6C, 0x73,
      0x00, 0x63, 0x68, 0x6C, 0x69, 0x73, 0x74, 0x00, 0x49, 0x00, 0x00, 0x00, 0x41, 0x00, 0x02, 0x00,
      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x42, 0x00,
      0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00,
      0x47, 0x00, 0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x01, 0x00,
      0x00, 0x00, 0x52, 0x00, 0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00,
      0x01, 0x00, 0x00, 0x00, 0x00, 0x63, 0x68, 0x72, 0x6F, 0x6D, 0x61, 0x74, 0x69, 0x63, 0x69, 0x74,
      0x69, 0x65, 0x73, 0x00, 0x63, 0x68, 0x72, 0x6F, 0x6D, 0x61, 0x74, 0x69, 0x63, 0x69, 0x74, 0x69,
      0x65, 0x73, 0x00, 0x20, 0x00, 0x00, 0x00, 0xF4, 0xD6, 0x23, 0x3F, 0x17, 0xF7, 0xA8, 0x3E, 0x19,
      0x9A, 0x99, 0x3E, 0xD2, 0x99, 0x19, 0x3F, 0x23, 0x9A, 0x19, 0x3E, 0xA1, 0xBF, 0x75, 0x3D, 0x37,
      0x1A, 0xA0, 0x3E, 0xB0, 0x72, 0xA8, 0x3E, 0x63, 0x6F, 0x6D, 0x70, 0x72, 0x65, 0x73, 0x73, 0x69,
      0x6F, 0x6E, 0x00, 0x63, 0x6F, 0x6D, 0x70, 0x72, 0x65, 0x73, 0x73, 0x69, 0x6F, 0x6E, 0x00, 0x01,
      0x00, 0x00, 0x00, 0x03, 0x64, 0x61, 0x74, 0x61, 0x57, 0x69, 0x6E, 0x64, 0x6F, 0x77, 0x00, 0x62,
      0x6F, 0x78, 0x32, 0x69, 0x00, 0x10, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
      0x00, 0x01, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x64, 0x69, 0x73, 0x70, 0x6C, 0x61, 0x79,
      0x57, 0x69, 0x6E, 0x64, 0x6F, 0x77, 0x00, 0x62, 0x6F, 0x78, 0x32, 0x69, 0x00, 0x10, 0x00, 0x00,
      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00,
      0x00, 0x6C, 0x69, 0x6E, 0x65, 0x4F, 0x72, 0x64, 0x65, 0x72, 0x00, 0x6C, 0x69, 0x6E, 0x65, 0x4F,
      0x72, 0x64, 0x65, 0x72, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x70, 0x69, 0x78, 0x65, 0x6C, 0x41,
      0x73, 0x70, 0x65, 0x63, 0x74, 0x52, 0x61, 0x74, 0x69, 0x6F, 0x00, 0x66, 0x6C, 0x6F, 0x61, 0x74,
      0x00, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00, 0x80, 0x3F, 0x73, 0x63, 0x72, 0x65, 0x65, 0x6E, 0x57,
      0x69, 0x6E, 0x64, 0x6F, 0x77, 0x43, 0x65, 0x6E, 0x74, 0x65, 0x72, 0x00, 0x76, 0x32, 0x66, 0x00,
      0x08, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x73, 0x63, 0x72, 0x65,
      0x65, 0x6E, 0x57, 0x69, 0x6E, 0x64, 0x6F, 0x77, 0x57, 0x69, 0x64, 0x74, 0x68, 0x00, 0x66, 0x6C,
      0x6F, 0x61, 0x74, 0x00, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00, 0x80, 0x3F, 0x00, 0x95, 0x01, 0x00,
      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x2B, 0x00, 0x00, 0x00, 0x78, 0x9C, 0x63,
      0x68, 0x68, 0x60, 0x60, 0x80, 0x60, 0x08, 0x60, 0x60, 0x38, 0xE0, 0x00, 0x13, 0x01, 0xB1, 0xF7,
      0x3B, 0xEE, 0x73, 0x6A, 0x68, 0xD8, 0xE7, 0xB4, 0xDF, 0x11, 0x22, 0xBF, 0xDF, 0x11, 0x04, 0x41,
      0x34, 0x98, 0x0D, 0x00, 0xDC, 0x52, 0x19, 0xC0, 0x0B, 0x64, 0xFD, 0x4E, 0xE0, 0x01, 0x7B, 0x61,
      0x1A, 0xC0
    };

    var texture = AsyncImageLoader.CreateFromImage(data);
    var mipmapLevel0 = texture.GetPixelData<float>(0);
    var mipmapLevel1 = texture.GetPixelData<float>(1);

    Assert.That(texture.format, Is.EqualTo(TextureFormat.RGBAFloat));
    Assert.That(texture.width, Is.EqualTo(2));
    Assert.That(texture.height, Is.EqualTo(2));
    Assert.That(texture.mipmapCount, Is.EqualTo(2));

    Assert.That(mipmapLevel0.Length, Is.EqualTo(4 * 4));
    Assert.That(mipmapLevel0, Is.EquivalentTo(new float[] {
      1.0f, 1.0f, 1.0f, 1.0f,
      .75f, .00f, .00f, .75f,
      .50f, .00f, .50f, .00f,
      .25f, .25f, .00f, .00f,
    }));

    Assert.That(mipmapLevel1.Length, Is.EqualTo(4));
    Assert.That(mipmapLevel1, Is.EquivalentTo(new float[] {
      .625f, .3125f, .375f, .4375f,
    }));
  }
}


Runtime/AsyncImageLoader/Pixels.cs
public partial class AsyncImageLoader {
  interface IPixel<TChannel> where TChannel : struct {
    static int ChannelCount() => throw new System.NotImplementedException();
    static int ChannelByteCount() => throw new System.NotImplementedException();
  }

  struct R16Pixel : IPixel<short> {
    public static int ChannelCount() => 1;
    public static int ChannelByteCount() => sizeof(short);

    public short r;
  }

  struct RFloatPixel : IPixel<float> {
    public static int ChannelCount() => 1;
    public static int ChannelByteCount() => sizeof(float);

    public float r;
  }

  struct RGB24Pixel : IPixel<byte> {
    public static int ChannelCount() => 3;
    public static int ChannelByteCount() => sizeof(byte);

    public byte r, g, b;
  }

  struct RGB48Pixel : IPixel<short> {
    public static int ChannelCount() => 3;
    public static int ChannelByteCount() => sizeof(short);

    public short r, g, b;
  }

  struct RGBA32Pixel : IPixel<byte> {
    public static int ChannelCount() => 4;
    public static int ChannelByteCount() => sizeof(byte);

    public byte r, g, b, a;
  }

  struct RGBA64Pixel : IPixel<short> {
    public static int ChannelCount() => 4;
    public static int ChannelByteCount() => sizeof(short);

    public short r, g, b, a;
  }

  struct RGBAFloatPixel : IPixel<float> {
    public static int ChannelCount() => 4;
    public static int ChannelByteCount() => sizeof(float);

    public float r, g, b, a;
  }
}


Runtime/AsyncImageLoader/FreeImage.cs
using System;
using System.Runtime.InteropServices;

public static partial class AsyncImageLoader {
  public static class FreeImage {
    public enum Format {
      FIF_UNKNOWN = -1,
      FIF_BMP = 0,
      FIF_ICO = 1,
      FIF_JPEG = 2,
      FIF_JNG = 3,
      FIF_KOALA = 4,
      FIF_LBM = 5,
      FIF_IFF = FIF_LBM,
      FIF_MNG = 6,
      FIF_PBM = 7,
      FIF_PBMRAW = 8,
      FIF_PCD = 9,
      FIF_PCX = 10,
      FIF_PGM = 11,
      FIF_PGMRAW = 12,
      FIF_PNG = 13,
      FIF_PPM = 14,
      FIF_PPMRAW = 15,
      FIF_RAS = 16,
      FIF_TARGA = 17,
      FIF_TIFF = 18,
      FIF_WBMP = 19,
      FIF_PSD = 20,
      FIF_CUT = 21,
      FIF_XBM = 22,
      FIF_XPM = 23,
      FIF_DDS = 24,
      FIF_GIF = 25,
      FIF_HDR = 26,
      FIF_FAXG3 = 27,
      FIF_SGI = 28,
      FIF_EXR = 29,
      FIF_J2K = 30,
      FIF_JP2 = 31,
      FIF_PFM = 32,
      FIF_PICT = 33,
      FIF_RAW = 34,
      FIF_WEBP = 35,
      FIF_JXR = 36
    }

    internal enum Type {
      /// <summary>
      /// unknown type
      /// </summary>
      FIT_UNKNOWN = 0,
      /// <summary>
      /// standard image : 1-, 4-, 8-, 16-, 24-, 32-bit
      /// </summary>
      FIT_BITMAP = 1,
      /// <summary>
      /// array of unsigned short : unsigned 16-bit
      /// </summary>
      FIT_UINT16 = 2,
      /// <summary>
      /// array of short : signed 16-bit
      /// </summary>
      FIT_INT16 = 3,
      /// <summary>
      /// array of unsigned long : unsigned 32-bit
      /// </summary>
      FIT_UINT32 = 4,
      /// <summary>
      /// array of long : signed 32-bit
      /// </summary>
      FIT_INT32 = 5,
      /// <summary>
      /// array of float : 32-bit IEEE floating point
      /// </summary>
      FIT_FLOAT = 6,
      /// <summary>
      /// array of double : 64-bit IEEE floating point
      /// </summary>
      FIT_DOUBLE = 7,
      /// <summary>
      /// array of FICOMPLEX : 2 x 64-bit IEEE floating point
      /// </summary>
      FIT_COMPLEX = 8,
      /// <summary>
      /// 48-bit RGB image : 3 x 16-bit
      /// </summary>
      FIT_RGB16 = 9,
      /// <summary>
      /// 64-bit RGBA image : 4 x 16-bit
      /// </summary>
      FIT_RGBA16 = 10,
      /// <summary>
      /// 96-bit RGB float image : 3 x 32-bit IEEE floating point
      /// </summary>
      FIT_RGBF = 11,
      /// <summary>
      /// 128-bit RGBA float image : 4 x 32-bit IEEE floating point
      /// </summary>
      FIT_RGBAF = 12
    }

    [System.Flags]
    internal enum LoadFlags {
      /// <summary>
      /// Default option for all types.
      /// </summary>
      DEFAULT = 0,
      /// <summary>
      /// Load the image as a 256 color image with ununsed palette entries, if it's 16 or 2 color.
      /// </summary>
      GIF_LOAD256 = 1,
      /// <summary>
      /// 'Play' the GIF to generate each frame (as 32bpp) instead of returning raw frame data when loading.
      /// </summary>
      GIF_PLAYBACK = 2,
      /// <summary>
      /// Convert to 32bpp and create an alpha channel from the AND-mask when loading.
      /// </summary>
      ICO_MAKEALPHA = 1,
      /// <summary>
      /// Load the file as fast as possible, sacrificing some quality.
      /// </summary>
      JPEG_FAST = 0x0001,
      /// <summary>
      /// Load the file with the best quality, sacrificing some speed.
      /// </summary>
      JPEG_ACCURATE = 0x0002,
      /// <summary>
      /// Load separated CMYK "as is" (use | to combine with other load flags).
      /// </summary>
      JPEG_CMYK = 0x0004,
      /// <summary>
      /// Load and rotate according to Exif 'Orientation' tag if available.
      /// </summary>
      JPEG_EXIFROTATE = 0x0008,
      /// <summary>
      /// Load the bitmap sized 768 x 512.
      /// </summary>
      PCD_BASE = 1,
      /// <summary>
      /// Load the bitmap sized 384 x 256.
      /// </summary>
      PCD_BASEDIV4 = 2,
      /// <summary>
      /// Load the bitmap sized 192 x 128.
      /// </summary>
      PCD_BASEDIV16 = 3,
      /// <summary>
      /// Avoid gamma correction.
      /// </summary>
      PNG_IGNOREGAMMA = 1,
      /// <summary>
      /// If set the loader converts RGB555 and ARGB8888 -> RGB888.
      /// </summary>
      TARGA_LOAD_RGB888 = 1,
      /// <summary>
      /// Reads tags for separated CMYK.
      /// </summary>
      TIFF_CMYK = 0x0001,
      /// <summary>
      /// Tries to load the JPEG preview image, embedded in
      /// Exif Metadata or load the image as RGB 24-bit if no 
      /// preview image is available.
      /// </summary>
      RAW_PREVIEW = 0x1,
      /// <summary>
      /// Loads the image as RGB 24-bit.
      /// </summary>
      RAW_DISPLAY = 0x2,
    }

    const string FreeImageLibrary = "FreeImage";

    [DllImport(FreeImageLibrary, EntryPoint = "FreeImage_IsLittleEndian")]
    internal static extern bool IsLittleEndian();

    [DllImport(FreeImageLibrary, EntryPoint = "FreeImage_GetFileTypeFromMemory")]
    internal static extern Format GetFileTypeFromMemory(IntPtr memory, int size);

    [DllImport(FreeImageLibrary, EntryPoint = "FreeImage_OpenMemory")]
    internal static extern IntPtr OpenMemory(IntPtr data, uint size_in_bytes);

    [DllImport(FreeImageLibrary, EntryPoint = "FreeImage_CloseMemory")]
    internal static extern IntPtr CloseMemory(IntPtr data);

    [DllImport(FreeImageLibrary, EntryPoint = "FreeImage_LoadFromMemory")]
    internal static extern IntPtr LoadFromMemory(Format format, IntPtr stream, int flags);

    [DllImport(FreeImageLibrary, EntryPoint = "FreeImage_Unload")]
    internal static extern void Unload(IntPtr dib);

    [DllImport(FreeImageLibrary, EntryPoint = "FreeImage_ConvertToRawBits")]
    internal static extern void ConvertToRawBits(IntPtr bits, IntPtr dib, int pitch, uint bpp, uint red_mask, uint green_mask, uint blue_mask, bool topdown);

    [DllImport(FreeImageLibrary, EntryPoint = "FreeImage_GetWidth")]
    internal static extern uint GetWidth(IntPtr handle);

    [DllImport(FreeImageLibrary, EntryPoint = "FreeImage_GetHeight")]
    internal static extern uint GetHeight(IntPtr handle);

    [DllImport(FreeImageLibrary, EntryPoint = "FreeImage_GetImageType")]
    internal static extern Type GetImageType(IntPtr dib);

    [DllImport(FreeImageLibrary, EntryPoint = "FreeImage_GetBPP")]
    internal static extern int GetBPP(IntPtr dib);

    [DllImport(FreeImageLibrary, EntryPoint = "FreeImage_GetBits")]
    internal static extern IntPtr GetBits(IntPtr dib);

    [DllImport(FreeImageLibrary, EntryPoint = "FreeImage_GetLine")]
    internal static extern int GetLine(IntPtr dib);

    [DllImport(FreeImageLibrary, EntryPoint = "FreeImage_GetPitch")]
    internal static extern int GetPitch(IntPtr dib);
  }
}
